mypy/checker.py

```diff
@@ -2141,6 +2142,37 @@ def check_assignment(self, lvalue: Lvalue, rvalue: Expression, infer_lvalue_type
                     rvalue_type = remove_instance_last_known_values(rvalue_type)
                 self.infer_variable_type(inferred, lvalue, rvalue_type, rvalue)
 
+    def try_infer_partial_generic_type_from_assignment(self,
+                                                       lvalue: Lvalue,
+                                                       rvalue: Expression) -> None:
+        """Try to infer a precise type for partial generic type from assignment.
+
+        Example where this happens:
+
+            x = []
+            if foo():
+                x = [1]  # Infer List[int] as type of 'x'
+        """
+        if (isinstance(lvalue, NameExpr)
```

Comment:
Should this actually be a `NameExpr` or `RefExpr` is fine? I am thinking about `self.x = []`, please add tests if yes.

Reply:
Partial types are not properly supported with attributes in general, so this is consistent with how things work elsewhere.